dvc/repo/experiments/serialize.py

```diff
@@ -0,0 +1,154 @@
+import json
+from dataclasses import asdict, dataclass, field
+from datetime import datetime
+from typing import TYPE_CHECKING, Any, Callable, Dict, Optional
+
+from dvc.exceptions import DvcException
+from dvc.repo.metrics.show import _gather_metrics
+from dvc.repo.params.show import _gather_params
+from dvc.utils import onerror_collect, relpath
+
+if TYPE_CHECKING:
+    from dvc.repo import Repo
+
+
+class DeserializeError(DvcException):
+    pass
+
+
+class _ISOEncoder(json.JSONEncoder):
+    def default(self, o: object) -> Any:
+        if isinstance(o, datetime):
+            return o.isoformat()
+        return super().default(o)
+
+
+@dataclass(frozen=True)
+class SerializableExp:
```

Comment:
The nature of our JSON format is arbitrary. It might be easier to just save json format directly and return that directly.

Reply:
(I do like structure, but our format is arbitrary)

Reply:
This is one of those places I'd like to move away from using json/yaml as our default serialization format anyways in favor of something that's not text based and is faster. Keeping it structured makes it easier to do that, and dealing with nested dicts of dicts everywhere in params/metrics/exp show is a lot harder to follow vs having proper dataclasses.

I am wondering if we want to start moving to `attrs` in core `dvc` though? We use attrs in some subprojects and dataclasses in some other subprojects and core dvc, but I wasn't sure if we ever made a final decision on using one vs the other.